secboot/secboot_tpm.go

```diff
@@ -244,3 +264,176 @@ func unlockEncryptedPartition(tpm *sb.TPMConnection, name, device, keyfile, pinf
 
 	return nil
 }
+
+func SealKey(key partition.EncryptionKey, params *SealKeyParams) error {
+	numModels := len(params.ModelParams)
+	if numModels < 1 {
+		return fmt.Errorf("at least one set of model-specific parameters is required")
+	}
+
+	tpm, err := sbConnectToDefaultTPM()
+	if err != nil {
+		return fmt.Errorf("cannot connect to TPM: %v", err)
+	}
+
+	modelPCRProfiles := make([]*sb.PCRProtectionProfile, 0, numModels)
+
+	for _, modelParams := range params.ModelParams {
+		modelProfile := sb.NewPCRProtectionProfile()
+
+		// Verify if all EFI image files exist
+		for _, chain := range modelParams.EFILoadChains {
+			if err := checkFilesPresence(chain); err != nil {
+				return err
+			}
+		}
+
+		// Add EFI secure boot policy profile
+		policyParams := sb.EFISecureBootPolicyProfileParams{
+			PCRAlgorithm:  tpm2.HashAlgorithmSHA256,
+			LoadSequences: buildLoadSequences(modelParams.EFILoadChains),
+			// TODO:UC20: set SignatureDbUpdateKeystore to support key rotation
+		}
+
+		if err := sbAddEFISecureBootPolicyProfile(modelProfile, &policyParams); err != nil {
+			return fmt.Errorf("cannot add EFI secure boot policy profile: %v", err)
+		}
+
+		// Add systemd EFI stub profile
+		if len(modelParams.KernelCmdlines) != 0 {
+			systemdStubParams := sb.SystemdEFIStubProfileParams{
+				PCRAlgorithm:   tpm2.HashAlgorithmSHA256,
+				PCRIndex:       tpmPCR,
+				KernelCmdlines: modelParams.KernelCmdlines,
+			}
+			if err := sbAddSystemdEFIStubProfile(modelProfile, &systemdStubParams); err != nil {
+				return fmt.Errorf("cannot add systemd EFI stub profile: %v", err)
+			}
+		}
+
+		// Add snap model profile
+		if modelParams.Model != nil {
+			snapModelParams := sb.SnapModelProfileParams{
+				PCRAlgorithm: tpm2.HashAlgorithmSHA256,
+				PCRIndex:     tpmPCR,
+				Models:       []*asserts.Model{modelParams.Model},
+			}
+			if err := sbAddSnapModelProfile(modelProfile, &snapModelParams); err != nil {
+				return fmt.Errorf("cannot add snap model profile: %v", err)
+			}
+		}
+
+		modelPCRProfiles = append(modelPCRProfiles, modelProfile)
+	}
+
+	var pcrProfile *sb.PCRProtectionProfile
+	if numModels > 1 {
+		pcrProfile = sb.NewPCRProtectionProfile().AddProfileOR(modelPCRProfiles...)
+	} else {
+		pcrProfile = modelPCRProfiles[0]
+	}
+
+	// Provision the TPM as late as possible
+	if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
+		return err
+	}
+
+	// If we are re-installing, there will be an a sealed keyfile that already
```

Comment:
What was the rationale for having the keyfile in the seed partition, rather than the boot partition? (I assume boot has been recreated at this point?)

Reply:
mostly, less write churn on seed than boot, it might be moot though if we need to reseal at each kernel refresh

Reply:
This was moved in f25b56c after a discussion, maybe @pedronis or @mvo can remember the exact reason?